### PR TITLE
Pokemon item rework

### PR DESCRIPTION
**Pull request by Ellamental2**
image

### CODE REVIEW

**Comment by Ellamental2** [comment]
included changes to the item sheets and ability to "consume item"

does not yet work with pokemon. will change this PR to a draft

**Comment by Ellamental2** [comment]
image
consume item reduces the quantity of the item by 1 in the players inventory

**Comment by Ellamental2** [comment]
before:
image

after:
image

no effects are applied, just bookkeeping

**Comment by Ellamental2** [comment]
Todo:
- [ ] consumable: false to the items in the template
- [ ] change items that are consumable to consumable: true.
- [ ] in the chat item.hbs add {{#if consumable}} around the button block so this doesn't show for unconsumable items (eg black belt, pokedex etc)
- [ ] change the item-items.hbs to show if it is consumable probably with a checkbox

File: module/sidebar/components/items-component.js [hunk 1]
```diff
@@ -16,30 +16,56 @@ export default class ItemsComponent extends Component {
...
 
         this.element.children(".item").on("click", function(event) {
-            const {itemId, itemOwner} = event.currentTarget.dataset;
-            game.actors.get(itemOwner).items.get(itemId).sheet._toChat(); 
+            let {itemId, itemOwner} = event.currentTarget.dataset;
+
+            if(game.actors.get(itemOwner).type == 'character') {
+                game.actors.get(itemOwner).items.get(itemId).sheet._toChat();
+            } else if (game.actors.get(itemOwner).type == 'pokemon') {
+                game.ptu.data.items.find(i => i.id == itemId).sheet._toChat();
```
Comment (dylanpiera) [inline module/sidebar/components/items-component.js:66]:
```suggestion
                game.ptu.data.items.find(i => i.id == itemId).sheet._toChat(itemOwner);
```

File: module/sidebar/components/items-component.js [hunk 2]
```diff
@@ -16,30 +16,56 @@ export default class ItemsComponent extends Component {
...
                 });
             }
         }
-        this.element.html(output);
+        if(output != "") { this.element.html(output); }
 
         this.element.children(".item").on("click", function(event) {
-            const {itemId, itemOwner} = event.currentTarget.dataset;
-            game.actors.get(itemOwner).items.get(itemId).sheet._toChat(); 
+            let {itemId, itemOwner} = event.currentTarget.dataset;
```
Comment (dylanpiera) [inline module/sidebar/components/items-component.js:61]:
```suggestion
            const {itemId, itemOwner} = event.currentTarget.dataset;
```

File: module/ptu.js [hunk 3]
```diff
@@ -25,6 +25,7 @@ import { PrepareCustomSpecies, UpdateCustomSpeciesData } from './custom-species.
 import { InitCustomTypings as initCustomTypings, UpdateCustomTypings } from './custom-typings.js'
 import { ChangeLog } from './forms/changelog-form.js'
 import { applyDamageToTargets, undoDamageToTargets, newApplyDamageToTargets, handleApplicatorItem, TakeAction } from './combat/damage-calc-tools.js'
+import { useItem } from "./utils/useItems.js";
```
Comment (dylanpiera) [inline module/ptu.js:28]:
```suggestion

```
Unnecessary clutter in game.ptu.utils

File: module/ptu.js [hunk 4]
```diff
@@ -83,6 +84,9 @@ export const ptu = {
       handleApplicatorItem,
       takeAction: TakeAction,
     },
+    useItems: {
+      useItem,
+    },
```
Comment (dylanpiera) [inline module/ptu.js:89]:
```suggestion
    
```
Unnecessary clutter in game.ptu.utils

File: module/utils/useItems.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+export async function useItem(event){
+    //prevent the default action of the button
+    event.preventDefault();
+
+    // Get the item ID and name from the button element's data-item-id and data-item-name attributes
+    let itemId = event.currentTarget.dataset.item;
+    let itemName = event.currentTarget.dataset.itemName;
+    let parentId = `${event.currentTarget.dataset.itemParentid}`; //convert parentId to string
```
Comment (dylanpiera) [inline module/utils/useItems.js:14]:
```suggestion
    const itemId = event.currentTarget.dataset.item;
    const itemName = event.currentTarget.dataset.itemName;
    const parentId = ""+event.currentTarget.dataset.itemParentid; //convert parentId to string
```
Why the conversion to string... is it not already a string?
Also ""+var or var+"" is a faster cast.

File: module/utils/useItems.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    
+    //disable the button
+    event.currentTarget.disabled = true;
```
Comment (dylanpiera) [inline module/utils/useItems.js:17]: Should this not be done by default and only if the quantity goes to <0?

File: module/utils/useItems.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    
+    //log the event to the console for dev purposes
+    // console.log(event);
+    // console.log(itemId);
+    // console.log(itemName);
+    // console.log(parentId);
+
+    //find the actor with id parentId
+    let actor = game.actors.get(parentId);
```
Comment (dylanpiera) [inline module/utils/useItems.js:26]:
```suggestion
    const actor = game.actors.get(parentId);
```

File: module/utils/useItems.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    console.log(actor);
+    // if the user of the item is of type pokemon
+    if (actor.type == "pokemon"){
+        console.log(`Consuming item with ID ${itemId} and name ${itemName}`);
+        //change the held item to none
+        actor.update({"data.heldItem": "None"});
```
Comment (dylanpiera) [inline module/utils/useItems.js:32]:
```suggestion
        await actor.update({"data.heldItem": "None"});        
```
Always await async calls

Comment (dylanpiera) [reply module/utils/useItems.js:32]: Also should there not be a check that this item is still in fact, there?

just a simple if heldItem == name of item

Players like to break things after all.

File: module/utils/useItems.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        actor.update({"data.heldItem": "None"});        
+    }
+    if (actor.type == "character"){
+        if(actor.items.get(itemId).system.quantity < 1){
+            ui.notifications.error("You don't have any of this item left.");
+            return;
+        }
+        console.log(`Consuming item with ID ${itemId} and name ${itemName}`);
+        //reduce the number of this item that the character has by 1
+        actor.items.get(itemId).update({"system.quantity": actor.items.get(itemId).system.quantity - 1});
```
Comment (dylanpiera) [inline module/utils/useItems.js:41]:
```suggestion
        const item = actor.items.get(itemId);
        await item.update({"system.quantity": Number(duplicate(item.system.quantity)) - 1});
```
Always duplicate changes to the same field. Number cast for safety sake.

File: templates/chat/item.hbs [hunk 10]
```diff
@@ -104,7 +104,14 @@
                         </div>
                     </div>
                 {{/if}}
+                <div class="w-100 p-2" style="border-top:1px solid #00000080;/* padding:10px 5px */">
+                    <button class="reduce-item-count" data-item="{{item.id}}" data-item-name="{{item.name}}" data-item-parentId="{{item.parent._id}}" onclick="useItem(event)">
```
Comment (dylanpiera) [inline templates/chat/item.hbs:108]:
```suggestion
                    <button class="reduce-item-count" data-item="{{item.id}}" data-item-name="{{item.name}}" data-item-parentId="{{owner}}" onclick="useItem(event)">
```

File: templates/chat/item.hbs [hunk 11]
```diff
@@ -104,7 +104,14 @@
...
+                        Consume Item
+                    </button>
+                </div>
             </div>
         </div>
     </div>
+
+    <script src="../../utils/use-item.js"></script>
```
Comment (dylanpiera) [inline templates/chat/item.hbs:116]:
```suggestion
```
Do not inject scripts into foundry files like this. Instead create hooks.

File: templates/item/item-ability-sheet.hbs [hunk 12]
```diff
@@ -52,35 +63,27 @@
...
-                        </div>
-                        <div class="hgss-body">
-                            <input
-                                class="ml-2"
-                                name="system.frequency"
-                                type="text"
-                                value="{{system.frequency}}"
-                                placeholder=""
-                            />
-                        </div>
```
Comment (dylanpiera) [inline templates/item/item-ability-sheet.hbs:68]: Why is this removed? Frequency exists in abilities (f.e. Intimidate)

File: module/item/item-sheet.js [hunk 13]
```diff
@@ -30,7 +30,7 @@ export class PTUItemSheet extends ItemSheet {
 
 		// Alternatively, you could use the following return statement to do a
 		// unique item sheet by type, like `weapon-sheet.html`.
-		return `${path}/item-${this.item.type}-sheet.html`;
+		return `${path}/item-${this.item.type}-sheet.hbs`;
```
Comment (dylanpiera) [inline module/item/item-sheet.js:33]: Can't add it as a suggestion because there were no changes at that point in this file.... but:

At line 159 of this file, add a parameter ownerId, and pass it along during creation

```
	/**
	 * Handle To Chat call.
	 * @private
	 */
	_toChat(ownerId) {
		switch(this.object.type) {
			case "move":
				return sendMoveMessage({
					speaker: ChatMessage.getSpeaker({
						actor: this.actor
					}),
					name: this.object.name,
					move: this.object.system,
					templateType: 'details',
					owner: ownerId
				});
			default: 
				return sendItemMessage({
					speaker: ChatMessage.getSpeaker({
						actor: this.actor
					}),
					item: this.object
					owner: ownerId
				});
		}
	}
```

**Review by dylanpiera — CHANGES_REQUESTED** [review]

File: module/utils/useItems.js (hunk 6 above)
Comment (Ellamental2) [reply module/utils/useItems.js:17]: this is done so that the button isn't clicked twice and the amount reduced twice.

File: templates/item/item-ability-sheet.hbs (hunk 12 above)
Comment (Ellamental2) [reply templates/item/item-ability-sheet.hbs:68]: this shouldn't be removed i'll add it back in. not sure why it is

Comment (Ellamental2) [reply templates/item/item-ability-sheet.hbs:68]: added this back in

**Comment by Ellamental2** [comment]
Changes:

<details>
<summary>
non-consumable items no longer have a button to consume item:
</summary>

image

</details>

<details>
<summary>
consumable now shows in the items sheet
</summary>

image
image

</detail>

**Comment by Ellamental2** [comment]
<details>
<Summary>
added a pop-up that allows the GM to stop a player from throwing a pokeball, before all the calculations take place. I find that my players will sometimes accidentally click pokeball and target the wrong pokemon, or in the middle of a battle and rather than having to say no to the capture then delete the pokeball and re-add the pokemon to the scene it would be easier to just say no at the beginning.
</summary>

image

</details>

**Comment by Ellamental2** [comment]
I think this fork probably has enough changes in it so I'll give it a rest.

File: module/combat/effects/pokeball_effects.js [hunk 14]
```diff
@@ -293,6 +293,14 @@ export async function ThrowPokeball(thrower, target, pokeball) {
...
     if(targetToken.actor.type != "pokemon") return;
 
+    if(pokeball.system.quantity < 1){
+        ui.notifications.error("You don't have any of those left!");
+        return;
+    }
+
+    console.log(`Consuming item with ID ${pokeball._id} and name ${pokeball.name}`);
+    //reduce the number of balls that the character has by 1
+    pokeball.update({"system.quantity": Number(duplicate(pokeball.system.quantity)) - 1});
```
Comment (dylanpiera) [inline module/combat/effects/pokeball_effects.js:303]:
```suggestion
    ui.notifications.info("Removed 1 quantity from Pokeball in inventory.")
    pokeball.update({"system.quantity": Number(duplicate(pokeball.system.quantity)) - 1});
```

File: module/sidebar/components/menu-component.js [hunk 15]
```diff
@@ -103,11 +103,51 @@ export default class MenuComponent extends Component {
             }
         })
 
-        this.element.children('#menu-content').children('.pokeball-item').on("click", (event) => {
+        this.element.children('#menu-content').children('.pokeball-item').on("click", async(event) => {
             const { entityId, ballName, ownerId } = event.target.dataset;
 
             const owner = game.actors.get(ownerId);
-            game.ptu.utils.throwPokeball(owner, game.user?.targets?.first(), owner?.items.get(entityId));
+
```
Comment (dylanpiera) [inline module/sidebar/components/menu-component.js:110]: This will ask the player if they can throw the pokeball. Not the GM. You require the run-as-gm API for that.

I don't have the time to currently explaint that so won't be merging it in for the final v10 beta but can tackle it next week

**Review by dylanpiera — CHANGES_REQUESTED** [review]

**Review by dylanpiera — APPROVED** [review]
Checked in Discord
